### PR TITLE
SDK-211: Porting Notebook REST API to Java SDK

### PR DESCRIPTION
**Pull request by abhijitjaiswal**
Added Notebook REST API to Java SDK as per SDK-211 requested by Minesh Patel. Kindly review and let me know for changes.

### CODE REVIEW

File: src/main/java/com/qubole/qds/sdk/java/details/NotebookApiImpl.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        if (cluster_id != null) {
+            params.put("cluster_id", cluster_id);
+        }
+        RequestDetails requestDetails = new RequestDetails(params, RequestDetails.Method.PUT, null);
+        requestDetails.allowToBeRetried();
+        return new GenericInvokableBuilderImpl<Notebook>(client, requestDetails, Notebook.class, "notebooks", notebook_id);
+    }
+
+    @Override
+    public InvokableBuilder<Notebook> delete(String notebook_id)
```
Comment (amoghmargoor) [inline src/main/java/com/qubole/qds/sdk/java/details/NotebookApiImpl.java]: Is `NoteBook` return type for delete ? Can you check for other API's too what his the response and instantiate InvokeBuilder with right type parameter ?

Comment (abhijitjaiswal) [inline src/main/java/com/qubole/qds/sdk/java/details/NotebookApiImpl.java]: I have added NotebookResult class which handles the success messages for Delete, Bind and Configure as well Notebook objects for Create and Clone.

File: src/main/java/com/qubole/qds/sdk/java/api/NotebookAPI.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.qubole.qds.sdk.java.api;
+
+import com.qubole.qds.sdk.java.entities.Notebook;;
```
Comment (abhijitjaiswal) [inline src/main/java/com/qubole/qds/sdk/java/api/NotebookAPI.java]: Done

File: src/main/java/com/qubole/qds/sdk/java/details/NotebookApiImpl.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        if (cluster_id != null) {
+            params.put("cluster_id", cluster_id);
+        }
+        RequestDetails requestDetails = new RequestDetails(params, RequestDetails.Method.PUT, null);
+        requestDetails.allowToBeRetried();
+        return new GenericInvokableBuilderImpl<Notebook>(client, requestDetails, Notebook.class, "notebooks", cloned_from_notebook, "clone");
+    }
+
+    @Override
+    public InvokableBuilder<Notebook> bind_notebook_to_cluster(String cluster_id, String notebook_id)
```
Comment (abhijitjaiswal) [inline src/main/java/com/qubole/qds/sdk/java/details/NotebookApiImpl.java]: I have changed it to camel case at places, but for Notebook class, the constructor would expect the params same as the result.

File: src/test/java/com/qubole/qds/sdk/java/unittests/TestNotebook.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        expectedRequestData.put("language", language);
+        expectedRequestData.put("name", name);
+        expectedRequestData.put("tags", tags);
+        expectedRequestData.put("arguments", arguments);
+        expectedRequestData.put("note_id", notebook_id);
+        assertRequestDetails(invokeargs, "POST", "commands", expectedRequestData, null, CommandResponse.class);
+    }
+
+    //Below Test cases are to be executed by configuring API token in AbstractTest class
+    public void testExecuteCreateNotebook() throws Exception
```
Comment (abhijitjaiswal) [inline src/test/java/com/qubole/qds/sdk/java/unittests/TestNotebook.java]: Done, Added Notebook Example and used generic string. The user can do the changes there and use the same, need not use system paramters.

File: src/main/java/com/qubole/qds/sdk/java/api/CommandApi.java [hunk 5]
```diff
@@ -146,4 +146,11 @@
      * @return new builder
      */
     public InvokableBuilder<Response> cancel(String queryId);
+
+    /**
+     * Corresponds to http://docs.qubole.com/en/latest/quick-start-guide/running-spark-app.html
```
Comment (abhijitjaiswal) [inline src/main/java/com/qubole/qds/sdk/java/api/CommandApi.java]: Done, changed the doc link

File: src/main/java/com/qubole/qds/sdk/java/api/BaseCommand.java [hunk 6]
```diff
@@ -34,7 +34,8 @@
         PIG,
         SHELL,
         COMPOSITE,
-        SPARK
+        SPARK,
+        NOTEBOOk
```
Comment (abhijitjaiswal) [inline src/main/java/com/qubole/qds/sdk/java/api/BaseCommand.java]: Done

File: src/main/java/com/qubole/qds/sdk/java/details/NotebookApiImpl.java (hunk 3 above)
Comment (amoghmargoor) [inline src/main/java/com/qubole/qds/sdk/java/details/NotebookApiImpl.java]: Thats fine .. I meant only for API and Builder ... We unfortunately have not followed camelCase for Json POJO (can be done using right annotations though) so Notebook or NotebookResult is ok.

File: src/test/java/com/qubole/qds/sdk/java/unittests/TestNotebook.java (hunk 4 above)
Comment (amoghmargoor) [inline src/test/java/com/qubole/qds/sdk/java/unittests/TestNotebook.java]: As we are exposing examples as a maven project, customer generally expect it to just work by invoking main function with some system params (However we have not lived upto it completely). Hence, example requires few modifications. We can make a submission later and remove examples from this PR if time does not permit it (will raise required JIRA for it). Basically, we would like to create a notebook first (without cloning) and then bind it to a cluster, then clone it, run it and then delete both the created notebooks. So system params will only take API token, and cluster required.

**Comment by amoghmargoor** [comment]
@abhijitjaiswal I just had a comment about examples. Other changes LGTM.

**Comment by abhijitjaiswal** [comment]
@amoghmargoor the changes in the examples is done as per our discussion, could you please approve, merge and make a release for the same. Let me know if you have any more comments, happy to implement :)

**Pull request merged**